evdev/eventio.py

```diff
@@ -46,6 +46,52 @@ def read_loop(self):
             for event in self.read():
                 yield event
 
+    def read_loop_blockless(self):
+        '''
+        Enter a blockless :func:`select.select()` loop that yields input events endlessly.
+        '''
+        
+        # previous -> [sec, usec, type, code, val]
+        previous = [0, 0, 0, 0, 0]
+        
+        # flag for 0
+        nothing = 0
+        
+        '''
+        It is different from it's blocking counterpart because it does not stop the execution when the device is
+        idle, instead it replicates the previous event until another one happens. This is made possible by
+        making use of a custom :func: 'read_blockless()' function that doesn't raise 'BlockingIOError' if there
+        are no available events at the moment.
+        '''
+        
+        def read_blockless(previous1, previous2, previous3, previous4, previous5):
+            '''
+            Read and yield a single input event as an instance of :class:`InputEvent <evdev.events.InputEvent>`.
+
+            Return previous event as an instance of :class:`InputEvent <evdev.events.InputEvent>` if there are
+            no pending input events.
+            '''
+        
+            # event -> (sec, usec, type, code, val)
+            present_event = _input.device_read(self.fd)
+        
+            try:
+                if present_event[2] + present_event[3] + present_event[4] != nothing:
+                    yield InputEvent(*present_event)
+        
+                else:
+                    yield InputEvent(present_event[0], present_event[1], previous3, previous4, previous5)
+            except TypeError:
+                yield InputEvent(previous1, previous2, previous3, previous4, previous5)
+
+        while True:
+            # selecting device and setting timeout to 0 / nothing, thus allowing non-blocking feature
+            select.select([self.fd], [], [], nothing)
+        
+            for event in read_blockless(previous[0], previous[1], previous[2], previous[3], previous[4]):
+                previous[0], previous[1], previous[2], previous[3], previous[4] = event.usec, event.sec, event.type, event.code, event.value
+                yield event
```

Comment:
Don't transform the events into arrays please. Transform the result of device_read into an InputEvent as early as possible. This is not readable otherwise, sorry

Reply:
Where is the `TypeError` originating from? Instead of relying on exceptions as a way to handle your edge case, it would be better if you added explicit return-early guards to your function (https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html, https://medium.com/swlh/return-early-pattern-3d18a41bba8. Code should not provoke exceptions. Exceptions should be the exception.

Reply:
Again, thanks for the feedback. I'm committing some changes soon taking those precious advices into account.

> Where is the TypeError originating from? 

It was originating from a part of the code that was trying to yield present_event as InputEvent, but failed due to device_read returning None when reading an idle device earlier in the code. Thankfully, that is not present anymore as I followed your advice and added explicit return-early guards.

Reply:
And instead of having a `previous` array, use an InputEvent object there as well.

- Arrays are for dynamic lists that may grow in size.
- If you really need a fixed-size datastructure and want it array-like, use tuples.
- But here, use an InputEvent instead.

If someone saw this code snippet:

`def read_blockless(previous1, previous2, previous3, previous4, previous5):`

would they have any clue at all what `previous2` is? Use `previous.usec` instead.

Reply:
`def read_one_blockless(prev_sec, prev_usec, prev_type, prev_code, prev_value):` is already better

but `def read_one_blockless(previous):` would be even better. And then access `previous.sec` etc. only where you actually need it, at the last moment.

Reply:
You're right! I didn't think about this before. I'm commiting these changes now, thanks!